src/webots/vrml/WbProtoManager.cpp

```diff
@@ -773,56 +771,88 @@ void WbProtoManager::exportProto(const QString &path, int category) {
     WbLog::error(tr("Impossible to export PROTO '%1' as the source file cannot be read.").arg(protoName));
 }
 
-void WbProtoManager::declareExternProto(const QString &protoName, const QString &protoPath, bool ephemeral) {
+void WbProtoManager::declareExternProto(const QString &protoName, const QString &protoPath, int type) {
   for (int i = 0; i < mExternProto.size(); ++i) {
-    if (mExternProto[i]->name() == protoName)
-      return;  // declaration already present
+    if (mExternProto[i]->name() == protoName) {
+      if ((mExternProto[i]->type() == WbExternProto::INSTANTIATED && type == WbExternProto::EPHEMERAL) ||
+          (mExternProto[i]->type() == WbExternProto::EPHEMERAL && type == WbExternProto::INSTANTIATED)) {
+        mExternProto[i]->setType(WbExternProto::BOTH);
+        emit externProtoListChanged();
+      }
+      return;
+    }
   }
 
-  mExternProto.push_back(new WbExternProtoInfo(protoName, protoPath, ephemeral));
+  mExternProto.push_back(new WbExternProto(protoName, protoPath, type));
+  emit externProtoListChanged();
 }
 
-void WbProtoManager::removeExternProto(const QString &protoName, bool allowEphemeralRemoval) {
+void WbProtoManager::removeEphemeralExternProto(const QString &protoName) {
   for (int i = 0; i < mExternProto.size(); ++i) {
     if (mExternProto[i]->name() == protoName) {
-      if (!mExternProto[i]->isEphemeral() || (mExternProto[i]->isEphemeral() && allowEphemeralRemoval))
+      if (mExternProto[i]->type() == WbExternProto::BOTH)
+        mExternProto[i]->setType(WbExternProto::INSTANTIATED);
+      else if (mExternProto[i]->type() == WbExternProto::EPHEMERAL)
         mExternProto.remove(i);
+      else
+        assert(true);  // only ephemerals should be removed using this function, instanciated are removed on a save
```

Comment:
it's just to ensure the point should is not reached. This function has to be exclusively used to remove ephemerals from the list (i.e., using the ephemeral panel remove button), if anything else triggers it, it means either the logic changed or you're trying to do something that wasn't designed for. Together with the comment the hope is that it should indicate to the developer that something not intended is going on.

Reply:
The check is good but not how it is written. Does it work if compiled in `release` mode?
A much better way to write it would be to remove the else and instead explicitly check for:
```
assert(mExternProto[i]->type() == WbExternProto::BOTH || mExternProto[i]->type() == WbExternProto::EPHEMERAL)
```

Reply:
ah ops, I see what Yannick was saying now, what I intended to do is `assert(false)` so that it triggers no matter what. Either way you're right, your approach is better, I'll fix that and other instances like it in the other PR. Thank you